modules/fedimint-ln-common/src/api.rs

```diff
@@ -127,3 +141,52 @@ where
         }
     }
 }
+
+/// Filter out duplicate gateways. This is necessary because different guardians
+/// may have different TTLs for the same gateway, so two
+/// `LightningGatewayAnnouncement`s representing the same gateway registration
+/// may not be equal.
+fn filter_duplicate_gateways(
+    gateways: Vec<LightningGatewayAnnouncement>,
+) -> Vec<LightningGatewayAnnouncement> {
+    let gateways_by_gateway_id =
+        gateways
+            .into_iter()
+            .fold(std::collections::HashMap::new(), |mut acc, gateway| {
+                acc.entry(gateway.info.gateway_id)
+                    .or_insert_with(Vec::new)
+                    .push(gateway);
+                acc
+            });
+
+    // For each gateway, we may have multiple announcements with different settings
+    // and/or TTLs. We want to filter out duplicates in a way that doesn't allow a
+    // malicious guardian to override the caller's view of the gateways by
+    // returning a gateway with a shorter TTL. Instead, if we receive multiple
+    // announcements for the same gateway ID, we only filter out announcements
+    // that have the same settings, keeping the one with the longest TTL.
+    gateways_by_gateway_id
+        .into_values()
```

Comment:
A bit confused. Why do we create a hashmap above if we just throw away the keys and only use the values?

Reply:
The gateways are still grouped by the keys (values are `Vec`s of GWs with the same id), even though they aren't used directly.